sycl/doc/extensions/proposed/sycl_ext_codeplay_cuda_cluster_group.asciidoc

```diff
@@ -0,0 +1,413 @@
+
+
+== Overview
+
+CUDA compute capability (cc) 9.0 (sm_90 and above) devices introduces a new level in the
+thread hierarchy, called as thread block clusters, in CUDA terminology. A thread
+block cluster, is a collection of thread blocks (a work-group in SYCL
+terminology) that run concurrently. The work-groups which make up a cluster 
+have the ability to access one another's local memory, and can be synchronized. 
+This has various applications, convolutions, GEMMs and FFTs to name a few.
+
+This proposal introduces a SYCL API to expose these capabilities, and defines a 
+mechanism to launch a kernel with clusters enabled, access the cluster's various
+ranges and id's from the device code, atomics at the cluster level as well as 
+synchronize the cluster. This proposal also introduces a device aspect to check 
+if the SYCL device supports a cluster launch, and a device query to obtain the 
+maximum supported cluster size.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_CODEPLAY_CUDA_CLUSTER_GROUP` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not version-ed, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Extension to `enum class aspect`
+
+[source]
+----
+namespace sycl {
+enum class aspect {
+  ...
+  ext_codeplay_cuda_cluster_group
+}
+}
+----
+
+A device requires the `ext_codeplay_cuda_cluster_group` aspect to
+support launching a kernel with the `cluster_size` property defined in the
+following section.
+
+
+=== Launching a kernel with a `cluster_group`
+
+Because of the special scheduling guarantees associated with a cluster launch,
+the backend must know which kernel would be using this feature. Thus, this 
+proposal introduces a new launch property called as `cluster_size` that will
+contain the cluster size as a number of work-groups.
+ 
+[source,c++]
+----
+namespace sycl::ext::codeplay::experimental::cuda {
+/**
+* Dim Dimensionality of the launch
+* size sycl::range specifying the number of work-groups in the cluster
+  in each dimension.
+*/
+template <int Dim>
+struct cluster_size {
+  cluster_size(const sycl::range<Dim>& size);
+  sycl::range<Dim> get_cluster_size();
+  ...
+};
+using cluster_size_key = cluster_size;
+} // namespace sycl::ext::codeplay::experimental::cuda
+----
+
+The property list can the be constructed as follows - 
+
+[source,c++]
+----
+properties  cluster_launch_property{cluster_size({1, 2, 1})};
+----
+
+Note that the total number of work-groups in the kernel must be a multiple of
```

Comment:
We should use the new format for notes here:

```
[_Note:_ This is a note. _{endnote}_]
```

Reply:
done in [824af1e](https://github.com/intel/llvm/pull/13594/commits/824af1e30203f42612a8c13bbe4740c91aeb1a2f)

Though I believe it is not rendering correctly ?

Reply:
It looks right to me.  What looks wrong to you?

Reply:
This is how I see it - 
```
[Note: the total number of work-groups in the kernel must be a multiple of the cluster size in each dimension. — end note]
```

was this expected ? I had assumed it would be rendered like when one uses the `NOTE:` syntax
```
NOTE: this is the note
```

Reply:
I think this is correct, there's an example of how this renders in [another ext](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_enqueue_functions.asciidoc).

Reply:
That's the expected rendering.

Reply:
Oh I see
Thanks !